src/Analysis/Ast/Impl/Types/PythonClassType.Generics.cs

```diff
@@ -0,0 +1,368 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Python.Analysis.Specializations.Typing;
+using Microsoft.Python.Analysis.Utilities;
+using Microsoft.Python.Analysis.Values;
+using Microsoft.Python.Core;
+
+namespace Microsoft.Python.Analysis.Types {
+    internal partial class PythonClassType : PythonType, IPythonClassType, IGenericType, IEquatable<IPythonClassType> {
+        private bool _isGeneric;
+        private object _genericParameterLock = new object();
+        private Dictionary<string, PythonClassType> _specificTypeCache;
+        private Dictionary<IGenericTypeParameter, IPythonType> _genericParameters;
+        private IReadOnlyList<IGenericTypeParameter> _parameters = new List<IGenericTypeParameter>();
+        private ReentrancyGuard<IPythonClassType> _genericSpecializationGuard = new ReentrancyGuard<IPythonClassType>();
+        private ReentrancyGuard<IPythonClassType> _genericResolutionGuard = new ReentrancyGuard<IPythonClassType>();
+
+        #region IGenericType
+        /// <summary>
+        /// List of unfilled generic type parameters. Represented as entries in the GenericParameters dictionary 
+        /// that have both key and value as generic type parameters
+        /// e.g
+        /// {T, T}
+        /// Where T is a generic type parameter that needs to be filled in by the class
+        /// </summary>
+        public virtual IReadOnlyList<IGenericTypeParameter> Parameters => _parameters ?? Array.Empty<IGenericTypeParameter>();
+
+        public virtual bool IsGeneric => _isGeneric;
+
+        /// <summary>
+        /// A Class is generic if it has one or more unfilled generic type parameters or one of its bases is generic
+        /// </summary>
+        public void DecideGeneric() {
```

Comment:
Method does not seem to be part of `IGeneric`. Does it need to be public? You may want to move it out of the `#region`

Reply:
Yes I did some refactors and did that, just haven't pushed yet.